app/controllers/mixins/generic_show_mixin.rb

```diff
@@ -24,8 +24,9 @@ def show
         if @record && @record.try(:ems_id)
           ems = ExtManagementSystem.find(@record.ems_id)
           tags = {:type => "node", :hostname => @record.name}.to_json
-          url = polymorphic_path(ems, :display => "ad_hoc_metrics", :tags => tags)
-          redirect_to(url)
+          redirect_to polymorphic_path(ems, :display => "ad_hoc_metrics", :tags => tags)
+        elsif self.class.respond_to?(:custom_display_modes) && self.class.custom_display_modes.index(@display)
+          custom_display_call(@display)
```

Comment:
Can you extract this "if responds ... custom_display_call..." to a function, and call that both here and from the `else` case on L41?